redis/src/main/scala/zio/redis/executor/cluster/RedisClusterExecutorLive.scala

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2021 John A. De Goes and the ZIO contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.redis.executor.cluster
+
+import zio.redis._
+import zio.redis.api.Cluster.AskingCommand
+import zio.redis.codec.StringUtf8Codec
+import zio.redis.executor.cluster.RedisClusterExecutorLive._
+import zio.redis.executor.node.RedisNodeExecutorLive
+import zio.redis.executor.{RedisConnectionLive, RedisExecutor}
+import zio.redis.options.Cluster._
+import zio.schema.codec.Codec
+import zio.{Chunk, Exit, IO, Ref, Schedule, Scope, UIO, ZIO, ZLayer, durationInt}
+
+import java.io.IOException
+import scala.util.Try
+
+private[redis] final case class RedisClusterExecutorLive(
+  clusterConnectionRef: Ref.Synchronized[ClusterConnection],
+  scope: Scope.Closeable
+) extends RedisExecutor {
+
+  def execute(command: Chunk[RespValue.BulkString]): IO[RedisError, RespValue] = {
+
+    def executeAsk(executorZIO: IO[RedisError.IOError, RedisExecutor], asking: Boolean = false) =
+      for {
+        executor <- executorZIO
+        _        <- ZIO.when(asking)(executor.execute(AskingCommand.resp((), StringUtf8Codec)))
+        res      <- executor.execute(command)
+      } yield res
+
+    def executeSafe(keySlot: Slot) = {
+      val recover = executeAsk(getExecutor(keySlot)).flatMap {
+        case e: RespValue.Error => ZIO.fail(e.toRedisError)
+        case success            => ZIO.succeed(success)
+      }.catchSome {
+        case e: RedisError.Ask   => executeAsk(getExecutor(e.address), asking = true)
+        case _: RedisError.Moved => refreshConnect() *> executeAsk(getExecutor(keySlot))
+      }
+      recover.retry(Scheduler)
+    }
+
+    for {
+      key    <- Try(command(1)).fold(_ => ZIO.fail(CusterKeyError), bs => ZIO.succeedNow(bs))
```

Comment:
Why not `ZIO.attempt`?

Reply:
Cause it can't fail) Why 'ZIO.attempt' ?)

Reply:
I am suggesting that you drop the Try and go through ZIO directly, and you are replying that it can't fail. If it can't fail, why are you using Try at all? :)

Reply:
Oh. I'm so stupid, sorry my bad. Don't understand that you are talking about Try, thanks. Ofcourse it is better to use attempt.

Reply:
All good :)